### PR TITLE
Firebase Authentication 

### PR DESCRIPTION
**Pull request by ChaAbby**
This ticket adds backend functionality to auth pages with Firebase backend.
Can now:

- When clicked forgot password button sends a request to get email with link to change password
-  Login button now log's user in through firebase
- Register now creates and saves a new account in firebase db
- Once user is registered they will get an email verification
- Only accounts with verified emails can sign in (Must change this when we have more ways to sign in)
- Toggle hidden password for login and register

[Notion Ticket](https://www.notion.so/Dev-Team-Backlog-61374ba6710c4e7ca03499d17c379425?p=6845491f38fd4468b9546f1d87ff9ad7&pm=s)

### CODE REVIEW

File: lib/views/login_register/login.dart [hunk 1]
```diff
@@ -93,28 +120,47 @@ class LoginScreen extends StatelessWidget {
...
+                                hintStyle: const TextStyle(color: Colors.grey),
                                 border: InputBorder.none,
+                                suffixIcon: IconButton(
+                                  icon: Icon(
+                                    _passwordVisible ? Icons.visibility : Icons.visibility_off,
+                                    color: Theme.of(context).primaryColorDark,
+                                  ),
+                                  onPressed: () {
+                                    setState(() {
+                                      _passwordVisible = !_passwordVisible;
```
Comment (ChaAbby) [inline lib/views/login_register/login.dart:158]: Password visible toggle functionality

File: lib/views/login_register/login.dart [hunk 2]
```diff
@@ -127,7 +173,56 @@ class LoginScreen extends StatelessWidget {
...
+                      onPressed: () async {
+                        final email = _email.text;
+                        final password = _password.text;
+                        try {
+                          await FirebaseAuth.instance.signInWithEmailAndPassword(
+                            email: email,
+                            password: password,
+                          );
+                          final user = FirebaseAuth.instance.currentUser;
+                          if (user?.emailVerified ?? false) {
```
Comment (ChaAbby) [inline lib/views/login_register/login.dart:185]: Sign in only if email verified user

File: lib/views/login_register/forgot_password.dart [hunk 3]
```diff
@@ -43,20 +63,49 @@ class ForgotPasswordScreen extends StatelessWidget {
...
                   ),
                 ),
                 Padding(
                   padding: const EdgeInsets.all(30),
                   child: BigButton(
-                    onPressed: () => navigator.push(const PasswordResetSentScreen()),
+                    onPressed: () async {
+                      final email = _email.text;
+                      try {
+                        await FirebaseAuth.instance.sendPasswordResetEmail(email: email);
```
Comment (ChaAbby) [inline lib/views/login_register/forgot_password.dart:85]: Currently all FirebaseAuth calls that involve a url link do not redirect back to the app. From my research it is common to use dynamic links in firebase for these redirects to the app. But google is going to deprecate this next year so it is not recommended. Are we okay with leaving it the way it is for now or should we look into alternatives to redirect to the app after clicking email link?

**Review by nate-thegrate — COMMENTED** [review]
Holy cow, that was fast—and it works!

I'm super impressed, and super looking forward to learning how to integrate Firebase into more of the app.

As per usual, I made [another pull request](https://github.com/theheartcenter-one/thc/pull/27) with feedback and refactoring and whatnot :)

**Review by nate-thegrate — APPROVED** [review]
Thank you!

I'm super pumped to get this merged in, but I'll wait for a bit in case @nickollas-johann wants a chance to review.

**Review by nickollas-johann — APPROVED** [review]
Good job!

**Pull request merged**